src/transformStateWithClones.js

```
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArray = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    const newState = { ...currentState };
+
+    switch (action.type) {
+      case 'addProperties':
+        for (const key of Object.keys(action.extraData)) {
+          newState[key] = action.extraData[key];
+        }
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      case 'clear':
+        for (const key of Object.keys(newState)) {
+          delete newState[key];
+        }
+        break;
+    }
```

It's a good practice to include a `default` case in a `switch` statement to handle any unexpected action types. This improves the robustness of your code. For instance, you could throw an error if an unknown action type is encountered.